common : introduce composable PEG parser combinators for chat parsing

aldehir opened the pull request:
Supporting new models requires implementing several features:

- Lazy grammar for tool calling (`tool_choice = auto`)
- Full grammar for forced tool calls and `response_format` (reasoning models)
- Parallel tool calls support
- Parsing of reasoning and tool call outputs

For reasoning models, the grammar must include reasoning or performance degrades significantly.

The real challenge is that each model uses a different output format:

- Harmony response output (gpt-oss)
- XML with typed parameters (Qwen3-Coder, MiniMax M2)
  - These models expect string arguments as raw content rather than JSON, which requires type awareness at parse time.
- Pseudo-function call (LFM2 e.g. `[get_weather(location="..."), ...]`)

Currently, the grammar and parsing exist as separate functions, which works but feels a bit fragile. I believe we can unify the two by using parser combinators to compose a PEG parser. That way the grammar definition becomes the parser.

## Proposed Solution

This PR introduces a generic PEG (Parsing Expression Grammar) parser to the common library, along with chat-specific extensions and a complete reference implementation for Qwen3-Coder.

I've noticed there's often a lag between when a model is supported by llama.cpp and when proper tool calling is fully implemented. This parser aims to close that gap by letting you define the grammar and parser at the same time, making it easier to add full tool calling support for new models.

### Parsing Expression Grammars (PEG)

PEG parsers are straightforward to implement as recursive descent parsers. While recursive descent parsers are known for backtracking, the majority of model output can be parsed with minimal backtracking, making them practical for this use case.

Parser combinators allow us to compose complex parsers from simple, reusable building blocks. This creates a DSL that closely mimics the grammar itself.

Rather than defining both a grammar and parsing function, we can build a PEG parser that generates a compatible GBNF grammar (with exceptions) and parses model output.

### Features

- **Partial parsing** for streaming input
- **Built-in JSON parsers** for common patterns
- **Grammar generation** for generating compatible GBNF grammars
- **AST generation** with semantic tags for structured extraction
- **Three common AST shapes** covering most model formats:
    - `simple` - Content with optional reasoning
    - `native` - Tool arguments as JSON objects
    - `constructed` - Tool arguments as separate entities (XML or pseudo-functions)

### Examples

<details>
<summary>Parser for models that emit tool arguments as JSON</summary>

```cpp
auto parser = build_chat_peg_native_parser([&](common_chat_peg_native_builder & p) {
    // Build choice of available tools
    auto tool_choice = p.choice();
    for (const auto & tool : tools) {
        const auto & function = tool.at("function");
        std::string name = function.at("name");
        const auto & schema = function.at("parameters");

        auto tool_name = p.json_member("name", "\"" + p.literal(name) + "\"");
        auto tool_args = p.json_member("arguments", p.schema(p.json(), "tool-" + name + "-schema", schema));

        tool_choice |= p.rule("tool-" + name, "{" << tool_name << "," << tool_args << "}");
    }

    // Define tool call structure
    auto tool_call = p.trigger_rule("tool-call",
        p.sequence({
            p.literal("<tool_call>["),
            tool_choice,
            p.literal("]</tool_call>")
        })
    );

    return p.sequence({
        p.content(p.until("<tool_call>")),
        p.optional(tool_call),
        p.end()
    });
});
```
</details>

<details>
<summary>Parser for models that emit XML tags for each argument</summary>

```cpp
auto parser = build_chat_peg_constructed_parser([&](common_chat_peg_constructed_builder & p) {
    auto location_arg = p.tool_arg(
        p.tool_arg_open("<parameter name=\"" + p.tool_arg_name(p.literal("location")) + "\">"),
        p.tool_arg_string_value(p.until("</parameter>")),
        p.tool_arg_close(p.literal("</parameter>"))
    );

    auto get_weather_tool = p.tool(p.sequence({
        p.tool_open("<function name=\"" + p.tool_name(p.literal("get_weather")) + "\">"),
        location_arg,
        p.tool_close(p.literal("</function>"))
    }));

    return p.sequence({
        p.content(p.until("<tool_call>")),
        p.literal("<tool_call>"),
        get_weather_tool,
        p.literal("</tool_call>"),
        p.end()
    });
});
```
</details>

<details>
<summary>Grammar generation</summary>

```cpp
data.grammar = build_grammar([&](const common_grammar_builder & builder) {
    foreach_function(params.tools, [&](const json & fn) {
        builder.resolve_refs(fn.at("parameters"));
    });
    parser.build_grammar(builder, data.grammar_lazy);
});
```
</details>

## Implementation Details

The PEG parsers are implemented using `std::variant` rather than traditional inheritance. This reduces boilerplate and leverages `std::visit` for type-safety. I initially had an OOP implementation, but it started becoming quite cumbersome and this seems like the lesser evil of the two.

```cpp
using common_peg_parser_variant = std::variant<
    common_peg_epsilon_parser,
    common_peg_start_parser,
    common_peg_end_parser,
    common_peg_literal_parser,
    common_peg_sequence_parser,
    common_peg_choice_parser,
    common_peg_repetition_parser,
    common_peg_and_parser,
    common_peg_not_parser,
    common_peg_any_parser,
    common_peg_space_parser,
    common_peg_chars_parser,
    common_peg_json_string_parser,
    common_peg_until_parser,
    common_peg_schema_parser,
    common_peg_rule_parser,
    common_peg_ref_parser,
    common_peg_atomic_parser,
    common_peg_tag_parser
>;
```

Both parsers and AST nodes are allocated in arena structures to minimize memory allocations.

```cpp
class common_peg_arena {
    std::vector<common_peg_parser_variant> parsers_;
    std::unordered_map<std::string, common_peg_parser_id> rules_;
    common_peg_parser_id root_ = COMMON_PEG_INVALID_PARSER_ID;
    ...

class common_peg_ast_arena {
    std::vector<common_peg_ast_node> nodes_;
    ...
```

Each parser variant is wrapped in a `common_peg_parser` value type to produce a DSL for composing parser combinators.

Parsers can return results `FAIL`, `SUCCESS`, or `NEED_MORE_INPUT`. This is how the partial parsing is implemented. It does not raise an exception on partial parse like `common/chat-parser.cpp`, because partial parses are still valid for streaming.

## Additional Changes

- Added `common_chat_peg_parse()` to `common/chat.cpp` and chat formats `COMMON_CHAT_FORMAT_PEG_(SIMPLE|NATIVE|CONSTRUCTED)` to support models parsed by a PEG parser.
  - The parser must be passed from chat param initialization to the parse function. To do this, I currently serialize the parser to JSON and then deserialize to `common_chat_syntax.parser`. I'm not a fan, but this seems the least intrusive method to integrate. I'll implement any alternative mechanisms if desired.
- Added `common/unicode.{cpp,h}` derived from `src/unicode.{cpp,h}`. As I understand, we should not include headers from `src/`, so I had to copy the implementation. It does deviate by returning a result rather than raising an exception.

More comprehensive documentation is added in `docs/development/parsing.md`. The tests are also fairly comprehensive, `tests/test-chat-peg-parser.cpp`.

---

I know this is a big PR. I tried to minimize the implementation, while keeping enough to demonstrate value. #15703 shows community desire for _something_ like this, although it doesn't have to be this implementation.

---

> [!NOTE]
> I wrote the majority of the parsing myself, as well as the complete examples in `test-chat-peg-parser.cpp`. AI assistance was used in some portions, notably:
>
> - Prototyping the interface to get an understanding for what works well.
> - Writing some of the unit tests, such as the basic ones and unicode.
> - The serialization/deserialization procedures were done mostly with AI.

aldehir committed:
common : implement parser combinators to simplify chat parsing

aldehir committed:
add virtual destructor to parser_base

aldehir committed:
fix memory leak from circular references of rules

aldehir committed:
implement gbnf grammar building

aldehir committed:
remove unused private variable

pwilkin commented:
Yes! This is exactly what I was thinking about :) can you give me push writes to your repo so I can contribute without doing PRs to PRs?

aldehir committed:
create a base visitor and implement id assignment as a visitor

aldehir committed:
fix const ref for grammar builder

aldehir committed:
clean up types, friend classes, and class declarations

aldehir committed:
remove builder usage from until_parser

aldehir committed:
Use a counter class to help assign rule ids

aldehir committed:
cache everything

aldehir committed:
add short description for each parser

aldehir committed:
create a type for the root parser

aldehir committed:
implement repetition parser

aldehir committed:
Make optional, one_or_more, and zero_or_more subclasses of repetition

aldehir committed:
improve context constructor

aldehir committed:
improve until parsing and add benchmarks

aldehir committed:
remove cached() pattern, cache in parser_base with specialized parsing functions for each parser

aldehir committed:
improve json parsing performance to better match legacy parsing

aldehir commented:
No rush. I am getting closer to a set of parsing functions that I'm happy with. The unfortunate part is I had to roll specialized parsers to maintain comparable performance with the existing parsing. A lexer would likely help, but optimized parsers for certain use cases is enough for now.

I added a benchmark in the test that implements the Command R2B parser, and compares it to the existing one. It seemed like a good one to illustrate.

```
// Benchmarks are over 100 iterations
Reasoning + Content:
   New parser avg: 23 us
Legacy parser avg: 450 us

Reasoning + Tool Call:
   New parser avg: 263 us
Legacy parser avg: 151 us
```

The existing parsing has a leg up with JSON. That said, it's still a fraction of a millisecond for a full prompt. I think most of the cost will go into the constrained decoding anyway. I'll have to benchmark larger JSON documents. Worst case, we can fall back to the implementation in `json-partial.cpp`. The intent here is to better support streaming JSON.

aldehir committed:
fix const auto * it for windows

aldehir committed:
move id assignment to classes instead of using a visitor

aldehir committed:
create named rules in the command r7b example

aldehir committed:
use '.' for any in GBNF

aldehir committed:
fix parens around choices in gbnf grammar

aldehir committed:
add convenience operators to turn strings to literals

aldehir committed:
add free-form operators for const char * to simplify defining literals

aldehir committed:
simplify test case parser

aldehir committed:
implement semantic actions

aldehir committed:
remove groups in favor of actions and a scratchpad

aldehir committed:
add built in actions for common operations

aldehir committed:
add actions to command r7b example

aldehir committed:
use std::default_searcher for platforms that don't have bm

aldehir committed:
improve parser_type handling and add cast helper

aldehir committed:
add partial result type to better control when to run actions

aldehir committed:
fix bug in until()

aldehir committed:
run actions on partial results by default

aldehir committed:
use common_chat_msg for result

aldehir committed:
add qwen3 example wip

aldehir committed:
trash partial idea and simplify

aldehir committed:
move action arguments to a struct

aldehir committed:
implement aho-corasick matcher for until_parser and to build exclusion grammars

aldehir committed:
use std::string for input, since std::string_view is incompatible with std::regex

pwilkin committed:
Refactor tests

aldehir committed:
improve qwen3 example

aldehir committed:
implement sax-style parsing and refactor

aldehir committed:
fix json string in test

aldehir committed:
rename classes to use common_chat_ prefix

aldehir committed:
remove is_ suffix from functions

aldehir committed:
rename from id_counter to just counter

pwilkin committed:
Final refactored tests

pwilkin commented on tests/test-chat-parser-combinator.cpp:
```diff
@@ -0,0 +1,1211 @@
+
+
+        return thinking + p.optional(p.space() + content) + p.zero_or_more(p.space() + tool_call);
+    });
```
Yeah, the quasi-XML structure is pretty painful due to the way how it mixes JSON and XML elements :/ it's not just Qwen-Coder, a lot of models recently have used it. 

I think the parser structure is fine as it is. One thing I believe might add to the modularity is we could have classes extending parser_builder with specific functionalities typical for various models and then call those methods if some patterns keep repeating (like `xml_tool_call_with_json_args()`).

pwilkin replied:
Yeah, the SAX-type / streaming parser looks good, it gives us more dynamic control over how to process things.

I definitely think once we've done a few examples, some of those patterns need their helpers, like this:

`"<think>" << p.add_rule("reasoning-content", p.until("</think>")) << "</think>"`

is going to appear in virtually every thinking model out there (just the tag might change).

pwilkin committed:
Fix executable name and editorconfig-checker

pwilkin committed:
Third time's the charm...

aldehir committed:
add trigger parser to begin lazy grammar rule generation

aldehir committed:
working lazy grammar

aldehir committed:
refactor json rules now that we check for reachability

aldehir committed:
reduce pointer usage

aldehir committed:
print out grammars in example

aldehir committed:
merge

aldehir commented on tests/CMakeLists.txt:
```diff
@@ -180,7 +180,19 @@ if (NOT WIN32 OR NOT BUILD_SHARED_LIBS)
+    combinator/*.cpp
+    combinator/*.hpp
+    test-chat-parser-combinator.cpp
+)
```
In my experience, globing files in Makefile generators leads to some annoying problems. I don't see this pattern used elsewhere in the code, so I much prefer to enumerate the files.

pwilkin replied:
True. I'll refactor the macro to list the files manually.

aldehir committed:
rename to chat-peg-parser* and common_chat_peg_parser*

aldehir committed:
starting unicode support

pwilkin commented on tests/test-grammar-integration.cpp:
```diff
@@ -46,14 +46,7 @@ static bool match_string(const std::string & input, llama_grammar * grammar) {
-    }
-
-    return false;
+    return std::any_of(stacks_cur.begin(), stacks_cur.end(), [](auto stack) { return stack.empty(); });
```
Ah yeah, was doing some followup on clang warnings and didn't realize this was in an unrelated file :)

pwilkin commented on tests/testcase.hpp:
Also, as @ggerganov mentioned the parser code part is basically unmaintained now, so it would be a good idea to introduce proper patterns from the get-go (including proper source file modularization).

pwilkin committed:
Revert unrelated changes

aldehir commented on tests/CMakeLists.txt:
```diff
@@ -180,6 +180,19 @@ if (NOT WIN32 OR NOT BUILD_SHARED_LIBS)
+target_link_libraries(test-chat-parser-combinator PRIVATE common)
+install(TARGETS test-chat-parser-combinator RUNTIME)
+add_test(NAME test-chat-parser-combinator COMMAND test-chat-parser-combinator)
+set_property(TEST test-chat-parser-combinator PROPERTY LABELS main)
```
I agree, I wasn't opposed to splitting it up into multiple files just the part about aggregating them to run in a single executable.

But adding support to link multiple source files to one executable works for me.

pwilkin committed:
New macros for CMakeLists to enable multi-file compilations

aldehir commented:
@pwilkin how about this?

```diff
diff --git a/tests/CMakeLists.txt b/tests/CMakeLists.txt
index 4850e16f7..e060aa55f 100644
--- a/tests/CMakeLists.txt
+++ b/tests/CMakeLists.txt
@@ -1,13 +1,15 @@
 llama_add_compile_flags()
 
 function(llama_build source)
+    set(TEST_SOURCES ${source} ${ARGN})
+
     if (DEFINED LLAMA_TEST_NAME)
         set(TEST_TARGET ${LLAMA_TEST_NAME})
     else()
         get_filename_component(TEST_TARGET ${source} NAME_WE)
     endif()
 
-    add_executable(${TEST_TARGET} ${source})
+    add_executable(${TEST_TARGET} ${TEST_SOURCES})
     target_link_libraries(${TEST_TARGET} PRIVATE common)
     install(TARGETS ${TEST_TARGET} RUNTIME)
 endfunction()
@@ -83,6 +85,8 @@ function(llama_build_and_test source)
     set(multiValueArgs ARGS)
     cmake_parse_arguments(LLAMA_TEST "${options}" "${oneValueArgs}" "${multiValueArgs}" ${ARGN})
 
+    set(TEST_SOURCES ${source} ${LLAMA_TEST_UNPARSED_ARGUMENTS} get-model.cpp)
+
     if (NOT DEFINED LLAMA_TEST_LABEL)
         set(LLAMA_TEST_LABEL "main")
     endif()
@@ -95,7 +99,7 @@ function(llama_build_and_test source)
         get_filename_component(TEST_TARGET ${source} NAME_WE)
     endif()
 
-    add_executable(${TEST_TARGET} ${source} get-model.cpp)
+    add_executable(${TEST_TARGET} ${TEST_SOURCES})
     install(TARGETS ${TEST_TARGET} RUNTIME)
     target_link_libraries(${TEST_TARGET} PRIVATE common)
 
@@ -181,8 +185,8 @@ endif()
 
 llama_build_and_test(test-chat-parser.cpp)
 
-# Chat PEG parser tests (modular)
-file(GLOB_RECURSE CHAT_PEG_PARSER_TEST_SOURCES
+llama_build_and_test(
+    test-chat-peg-parser.cpp
     chat-peg-parser/simple_tokenizer.cpp
     chat-peg-parser/benchmark.cpp
     chat-peg-parser/test-actions.cpp
@@ -195,13 +199,7 @@ file(GLOB_RECURSE CHAT_PEG_PARSER_TEST_SOURCES
     chat-peg-parser/test-partial-parsing.cpp
     chat-peg-parser/test-recursive-references.cpp
     chat-peg-parser/tests.h
-    test-chat-peg-parser.cpp
 )
-add_executable(test-chat-peg-parser ${CHAT_PEG_PARSER_TEST_SOURCES})
-target_link_libraries(test-chat-peg-parser PRIVATE common)
-install(TARGETS test-chat-peg-parser RUNTIME)
-add_test(NAME test-chat-peg-parser COMMAND test-chat-peg-parser)
-set_property(TEST test-chat-peg-parser PROPERTY LABELS main)
 
 llama_build_and_test(test-chat-template.cpp)
 llama_build_and_test(test-json-partial.cpp)
```

aldehir committed:
add unicode support to char_parser

aldehir committed:
use unparsed args as additional sources

aldehir commented on tests/test-chat-peg-parser.cpp:
```diff
@@ -0,0 +1,32 @@
+    test_command7_parser_compare command7_compare_test;
+    command7_compare_test.run_comparison(100);
+
+    return 0;
```
Should return non-zero if a test fails, otherwise CI won't pick up a failing test!

I'm going to change this to the implementation I posted earlier, but I do want to keep your formatting. I'll keep test runner in `tests/` so it can be shared across the project (baring any objections).

pwilkin replied:
I'm doing the test reimplementation right now, so give me a moment :)

aldehir replied:
Ah, ok! No problem. I appreciate the help.

pwilkin committed:
Refactor tests to new harness

pwilkin committed:
Fix CMakeLists

aldehir commented:
Perfect!

One nit: you don't need to include .h in the sources. CMake will compile with `-MD` to obtain header dependencies for source files and include them in the dependency chain. That said, other sections in the code include .h so we can leave it for consistency.

aldehir committed:
fix rate calculation

aldehir committed:
add unicode tests

aldehir committed:
fix trailing whitespace and line endings

skip-checks: true

pwilkin committed:
Helpers + rewrite qwen3 with helpers

pwilkin committed:
Fix whitespace

pwilkin commented:
All right, I wrote a couple of helpers + refactored the Qwen3 example parser to use those helpers, let me know what you think - I believe those helpers will capture a lot of the typical implementations.

aldehir committed:
extract unicode functions to separate file

aldehir committed:
refactor parse unicode function

aldehir commented:
@hksdpc255 thanks! I absolutely agree with the pain points. I'm hoping this addresses most of them while not creating a new class of problems. I'm hoping to hook it up to an actual model by tomorrow and see how it goes. Right now it's just a set of tests.

aldehir committed:
fix compiler error

aldehir committed:
improve construction of sequence/choice parsers

aldehir committed:
be less clever

aldehir committed:
add make_parser helper function

aldehir committed:
expand usage of make_parser, alias common_chat_msg_peg_parser_builder to builder in source

aldehir committed:
lower bench iterations

aldehir committed:
add unicode support to until_parser

aldehir committed:
add unicode support to json_string_parser

aldehir committed:
clean up unicode tests

aldehir committed:
reduce unicode details to match src/unicode.cpp

aldehir committed:
simplify even further

aldehir committed:
remove unused functions

aldehir committed:
fix type

aldehir committed:
reformat char class parsing

aldehir committed:
clean up json string parser

aldehir committed:
clean up + fix diagnostics

aldehir committed:
reorder includes

aldehir committed:
compact builder functions

aldehir committed:
replace action_parser with capture_parser, rename env to semantics

aldehir committed:
rename env to semantics

aldehir committed:
clean up common_chat_parse_context

aldehir committed:
move type() to below constant

aldehir committed:
use default constructor for common_chat_peg_parser

aldehir committed:
make all operators functions for consistency

aldehir committed:
fix compilation errors in test-optional.cpp

aldehir committed:
simplify result values

aldehir committed:
rename json_string_unquoted to json_string_content

pwilkin committed:
Move helper to separate class, add separate explicit and helper classes

pwilkin committed:
Whitespace

pwilkin committed:
Change + to append()

pwilkin committed:
Reformat

pwilkin committed:
Add extra helpers, tests and Minimax example

pwilkin committed:
Add some extra optional debugging prints + real example of how to use them

pwilkin commented:
Aight, Minimax M2 and Seed-OSS are up. With the first one, I did a stupid mistake of doing different tool definitions from tool calls, so couldn't get a proper parse, so I added some debugging prints to go + a live example of how to use them :)

BTW, the current solution is if an *incorrect* function call is detected, it's still marked as a success since zero_or_more always succeeds, not sure if we don't want to pass a failure over somehow (as in, zero_or_more only trivially succeeds if the rest is empty?)

aldehir committed:
fix bug in repetitions when min_count = 0 reports failures

aldehir committed:
dump rule in debug

aldehir committed:
fix token accumulation and assert parsing never fails

aldehir commented on common/chat-peg-parser.cpp:
Well, that explains why your changes sometimes have trailing whitespace!

I'm a Neovim user, I use the clangd lsp server as well. It works really well, but I don't have the quickfix code action bound to a key sequence. I should do that, thanks!

aldehir commented on common/chat-peg-parser-helper.h:
```diff
@@ -0,0 +1,118 @@
+    if (ev.rule.find("arg-string") != std::string::npos && ev.ending() && ev.success()) {
+        auto & tc = semantics.tool_calls.back();
+        tc.arguments += "\"";
+    }
```
This was causing me some grief, because it matches the "arg-string-content" above and was adding two quotes to the end.

I don't think we should do a search here. I assume you mean to match `arg-string-<param>`. We can still produce those rules, but wrap them with an `arg-string` rule and use direct string comparison. Just to avoid tiny little bugs like this.

aldehir replied:
Actually, never mind. I see why you did that. Ok, I have to rethink this.

aldehir committed:
indent debug by depth

aldehir committed:
use LOG_* in tests so logs sync up with test logs

pwilkin committed:
- Add selective testing
- Refactor all messaging to use LOG_ERR
- Fix lack of argument / tool name capturing
- Temporary fix for double event capture

aldehir committed:
refactor rule() and introduce ref()

aldehir committed:
clean up visitor

aldehir committed:
clean up indirection in root parser w.r.t rules

aldehir committed:
replace aho-corasick automation with a simple trie

aldehir committed:
store shared ptr directly in parser classes

aldehir committed:
merge

aldehir commented:
Should probably revert the test logging, it'll be a hard sell if it breaks CI. We can wrap the debug parser output in a compilation flag temporarily.

tests are currently broken, something in the helper function for qwen3. I removed the newlines in the input during tool call because that's how their own parser expects the model output. I'll look at it later.

aldehir committed:
refactor to use value semantics with std::variant/std::visit

aldehir committed:
Reset prev for qwen3 helper example variant

aldehir committed:
simplify trie_matcher result

aldehir committed:
fix linting issues

aldehir committed:
add annotations to rules

aldehir committed:
revert test workaround

pwilkin commented:
Nah, GMT+1, I usually sleep in those hours ;)

aldehir committed:
implement serializing the parser

aldehir committed:
remove redundant parsers

aldehir committed:
remove tests

aldehir committed:
gbnf generation fixes

aldehir committed:
remove LOG_* use in tests

aldehir committed:
update gbnf tests to test entire grammar

aldehir committed:
clean up gbnf generation and fix a few bugs

aldehir committed:
fix typo in test output

aldehir committed:
remove implicit conversion rules

aldehir committed:
improve test output

aldehir committed:
rename trie_matcher to trie

aldehir committed:
simplify trie to just know if a node is the end of a word

aldehir committed:
remove common_chat_ prefix and ensure a common_peg_ prefix to all types

aldehir committed:
rename chat-peg-parser -> peg-parser

aldehir committed:
promote chat-peg-parser-helper to chat-peg-parser

aldehir committed:
checkpoint

aldehir committed:
use a static_assert to ensure we handle every branch

aldehir committed:
inline trivial peg parser builders

aldehir committed:
use json strings for now

aldehir committed:
implement basic and native chat peg parser builders/extractors

aldehir committed:
resolve refs to their rules

aldehir committed:
remove packrat caching (for now)

aldehir committed:
update tests

aldehir committed:
compare parsers with incremental input

aldehir committed:
Merge branch 'upstream/master' into parser-combinators

aldehir committed:
benchmark both complete and incremental parsing

aldehir committed:
add raw string generation from json schema

aldehir committed:
add support for string schemas in gbnf generation

aldehir committed:
fix qwen example to include \n

aldehir committed:
tidy up example

aldehir committed:
rename extractor to mapper

aldehir committed:
rename ast_arena to ast

aldehir committed:
place basic tests into one

aldehir committed:
use gbnf_format_literal from json-schema-to-grammar

aldehir committed:
integrate parser with common/chat and server

aldehir committed:
clean up schema and serialization

aldehir committed:
add json-schema raw string tests

aldehir committed:
clean up json creation and remove capture parser

aldehir committed:
trim spaces from reasoning and content

aldehir committed:
clean up redundant rules and comments

aldehir committed:
rename input_is_complete to is_partial to match rest of project

aldehir committed:
simplify json rules

aldehir committed:
remove extraneous file

aldehir committed:
remove comment

aldehir committed:
implement += and |= operators

aldehir committed:
add comments to qwen3 implementation

aldehir committed:
reorder arguments to common_chat_peg_parse

aldehir committed:
remove commented outdated tests

aldehir committed:
add explicit copy constructor

aldehir committed:
fix operators and constness

aldehir committed:
wip: update test-chat for qwen3-coder

aldehir committed:
bring json parser closer to json-schema-to-grammar rules

aldehir committed:
trim trailing space for most things

aldehir committed:
fix qwen3 coder rules w.r.t. trailing spaces

aldehir committed:
group rules

aldehir committed:
do not trim trailing space from string args

aldehir committed:
tweak spacing of qwen3 grammar

aldehir committed:
update qwen3-coder tests

aldehir committed:
qwen3-coder small fixes

aldehir committed:
place parser in common_chat_syntax to simplify invocation

aldehir committed:
use std::set to collect rules to keep order predictable for tests

aldehir committed:
initialize parser to make certain platforms happy

aldehir committed:
revert back to std::unordered_set, sort rule names at the end instead

aldehir committed:
uncomment rest of chat tests

aldehir committed:
define explicit default constructor

aldehir committed:
merge with master

aldehir committed:
improve arena init and server integration

aldehir committed:
fix chat test

aldehir committed:
add json_member()

aldehir committed:
add a comprehensive native example

aldehir committed:
clean up example qwen test and add response_format example to native test

aldehir committed:
make build_peg_parser accept std::function instead of template

aldehir committed:
change peg parser parameters into const ref

aldehir committed:
push tool call on tool open for constructed parser

aldehir committed:
add parsing documentation

aldehir committed:
clean up some comments

aldehir committed:
add json schema support to qwen3-coder

aldehir committed:
add id initializer in tests

aldehir committed:
remove grammar debug line from qwen3-coder

aldehir committed:
refactor qwen3-coder to use sequence over operators

aldehir committed:
only call common_chat_peg_parse if appropriate format

aldehir commented:
Well, I think the implementation is fairly complete.

There are a few outstanding issues with the JSON schema to grammar for raw strings, but I think those should be addressed separately.

I'm going to do one last pass and extract the Qwen3-Coder implementation into a separate PR. This one has gotten pretty big, although half of it is tests.

aldehir committed:
simplify qwen3-coder space handling

aldehir committed:
revert qwen3-coder implementation

aldehir committed:
revert json-schema-to-grammar changes

aldehir committed:
remove unnecessary forward declaration

aldehir committed:
Merge branch 'master' into parser-combinators

aldehir commented:
I have pushed the Qwen3-Coder implementation to a separate branch in my repo: https://github.com/aldehir/llama.cpp/tree/qwen3-coder-peg-parser

The helpers and any testing recommendations should probably go in a separate PR with the first usage of this new parsing mechanism (if approved). Those tests would most likely impact `tests/test-chat.cpp`.

I don't view the existing mechanism as a buggy mess, but I do believe there is an opportunity for improvement by aligning the grammar and parsing.

aldehir commented:
@ggerganov @CISC @ngxson Please take a look. Curious whether you think something like this would be a good fit.

pwilkin commented:
@ngxson Yeah, that's what I'm trying to convince @aldehir of: if we replace the current approach with this PEG parser + helpers, we can hopefully not only have an easier way of adding new chat formats, but also a unified test framework for all supported templates to go with it.

ngxson commented:
> @ngxson Yeah, that's what I'm trying to convince @aldehir of: if we replace the current approach with this PEG parser + helpers, we can hopefully not only have an easier way of adding new chat formats, but also a unified test framework for all supported templates to go with it.

Replacing all regex-based parser with PEG seems to be a logical next step, what's current holding it back?

aldehir commented:
> what's current holding it back?

Nothing! I have already demonstrated use with Qwen3-Coder and it works really well.  Where @pwilkin and I have differing opinions is in what should be a utility function, but we'll iron that out as we go.

There are a few edge cases in the `json-schema-to-grammar` that do need addressing, and I am actively working those.

aldehir committed:
small adjustment to until_parser

aldehir committed:
Merge branch 'master' into parser-combinators

ggerganov reviewed (commented):
@aldehir Looks like nice work!

aldehir committed:
rename C/C++ files to use dashes

aldehir committed:
Merge branch 'master' into parser-combinators

aldehir committed:
codeowners : add aldehir to peg-parser and related files

aldehir commented:
> @aldehir Looks like nice work!

Thank you!

I added myself to CODEOWNERS for the related files. I also updated the description in light of the recent AI disclosure policy.

ggerganov reviewed (approved):
Great! Let's merge this if there are no additional comments @pwilkin @ngxson

ngxson reviewed (approved):
LGTM, good jobs @aldehir !

pwilkin commented:
Let's merge so we can start building upon it :) Kudos to @aldehir for all his hard work here.
